worksheet_B/Mandelbrot.cpp

```diff
@@ -1,56 +1,232 @@
+	rgbMax = rgb.r > rgb.g ? (rgb.r > rgb.b ? rgb.r : rgb.b) : (rgb.g > rgb.b ? rgb.g : rgb.b);
+
+	hsv.v = rgbMax;
+	if (hsv.v == 0)
+	{
+		hsv.h = 0;
+		hsv.s = 0;
+		return hsv;
+	}
+
+	hsv.s = 255 * long(rgbMax - rgbMin) / hsv.v;
+	if (hsv.s == 0)
+	{
+		hsv.h = 0;
+		return hsv;
+	}
+
+	if (rgbMax == rgb.r)
+		hsv.h = 0 + 43 * (rgb.g - rgb.b) / (rgbMax - rgbMin);
+	else if (rgbMax == rgb.g)
+		hsv.h = 85 + 43 * (rgb.b - rgb.r) / (rgbMax - rgbMin);
+	else
+		hsv.h = 171 + 43 * (rgb.r - rgb.g) / (rgbMax - rgbMin);
+
+	return hsv;
+}
+// Initialise the image
+CImg<unsigned char> image(width, height, 1, 3, 0);
+
+// Display the image
+CImgDisplay display(image, "Mandelbrot Set");
+
+// Minimum and maximum coordinates for the fractal
+double minX = -2, maxX = 1, minY = -1.5, maxY = 1.5;
+
+// The sum of the X coordinates the result will be 3;
+double maxXSum = maxX + minX * -1;
+
+// The amount to move one pixel on any axis because it is a 3x3 square Divide it by the amount of pixels which is 800
+double onePix = maxXSum / width;
+
+// Calculate the X an Y coordinate of the center of the screen Starting is currentXc=0.5, currentYc=0.0
+double currentXc = minX + width/2 * onePix;
+double currentYc = maxY - width/2 * onePix;
+
+// To have a referance of how many time I have zoomed in
+int zoomLevel = 0;
+
+// I put the image generation inside a function so I could run the function on mouse click.
+// Zoom is the level of zoom we want to divide by, mouseX and mouseY is the mouse position of the click or the center of the screen
+void generateImg(double zoom, double mouseX, double mouseY){
+	// divide by zoom to make the coordinates smaller
+	maxXSum = maxXSum/zoom;
+	
+	// Keep the coordinates of the initial click. I couldn't find a way to recalculate the coordinates after every click, because it would not center the next calculation
+	if (zoomLevel < 2) {
+		currentXc = minX + mouseX * onePix;
+		currentYc = minY + mouseY * onePix;
+	}
+	// Half of the maxSum to know how much to set min and max coordinates based on the currentXc and currentYc coordinates
+	double zoomOffsetSum = maxXSum / 2;
+	// Recalculate the X and Y coordinates
+	double minX = currentXc - zoomOffsetSum;
+	double maxX = currentXc + zoomOffsetSum;
+	double minY = currentYc - zoomOffsetSum;
+	double maxY = currentYc + zoomOffsetSum;
+
+
+	for (int pixelY = 0; pixelY < image.height(); pixelY++)
+	{
+		// Map the y coordinate into the range minY to maxY
+
+		double y0 = (pixelY / height * (maxY - minY) + minY);
+
+		
+		for (int pixelX = 0; pixelX < image.width(); pixelX++)
+		{
+			// Map the x coordinate into the range minX to maxX
+			double x0 = (pixelX / width * (maxX - minX) + minX);
+
+			
+			// Set the temp coordinate to x0 and y0
+			double xi = x0;
+			double yi = y0;
+			// Save the iteration outside of the loop
+			int i = 0;
+			// Iterate for 200 if it is within the mandle brot set
+			for (i = 0; i < 200 && xi*xi + yi*yi <= 4; i++)

```

Comment:
Would be better to have a constant for maximum iterations rather than hard-coding the value `200`